### PR TITLE
Improve Kruskal's MST by optimizing Union-Find

### LINKED ISSUE #841: Optimizing Kruskal's MST
The current implementation takes **O(E*V+E*logE)** time.

I suggest the following changes:-

- Optimize Union Find to reduce time per operation from **O(V)** to **O(1)** amortized.
- Replace the Min Heap with a sorted array to improve speed and memory.
- Reduce the memory used by mst vector from **E** to **V-1**

This will improve the runtime complexity to **O(E*logE)** and reduce memory by a good factor

### PR DESCRIPTION
**Pull request by SohamTamba**
Fixes #841 

**Summary :-**  

Replaced naive Disjoint-set data structure and search algorithm with a more optimal one that noticeably improves performance with a slight decrease in memory overhead.

**Benchmark Results:-**
|E| = 5 * |V|

For |V| = 1000, 

Old Implementation:              2.777 ms (17204 allocations: 790.56 KiB)
Optimized Implementation:   1.529 ms (17981 allocations: 771.64 KiB)

For |V| = 10000, 

Old Implementation:            164.897 ms (281392 allocations: 9.40 MiB)
Optimized Implementation:   28.465 ms (278255 allocations: 9.05 MiB)

For |V| = 100000, 

Old Implementation:                13.637 s (2957230 allocations: 92.42 MiB)
Optimized Implementation:   666.549 ms (3044722 allocations: 90.71 MiB)

### CODE REVIEW

**Comment by SohamTamba** [comment]
I also made an orphan [branch](https://github.com/SohamTamba/LightGraphs.jl/tree/kruskal_test) that contains the current implementation exposed as kruskal_mst_old and the proposed one as kruskal_mst. 

This branch also contains a benchmark [script](url) that one can run to try and reproduce the trend of results I have got. I'd appreciate if someone can verify my results on a different machine.

File: src/spanningtrees/kruskal.jl [hunk 1]
```diff
@@ -6,27 +6,40 @@ end
...
-        if vs[i] == pid
-            vs[i] = qid
-        end
+
+function find_set!(set_id::Vector{U}, q::U) where U<:Integer
+    
+    root = q
+    while root != set_id[root]
+        root = set_id[root]
+    end
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: Why is this a `while` statement as opposed to an `if`?

File: src/spanningtrees/kruskal.jl [hunk 2]
```diff
@@ -6,27 +6,40 @@ end
 isless(e1::KruskalHeapEntry, e2::KruskalHeapEntry) = e1.dist < e2.dist
 
 """
-    quick_find!(vs, p, q)
+    find_set!(vs, q)
 
 Perform [Quick-Find algorithm](https://en.wikipedia.org/wiki/Disjoint-set_data_structure)
-on a given pair of vertices `p`and `q`, and make a connection between them in the vector `vs`.
+on a given vertice `q`, finds the component id it belongs to.
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: "vertex" is the singular. Suggest
"Perform ... on vertex `q` and return the component id in `vs` to which it belongs."

File: src/spanningtrees/kruskal.jl [hunk 3]
```diff
@@ -6,27 +6,40 @@ end
...
 
+union_set!{U<:Integer}(set_id::Vector{U}, p::U, q::U) = (set_id[find_set!(set_id, p)] = find_set!(set_id, q))
+
 """
     kruskal_mst(g, distmx=weights(g))
 
 Return a vector of edges representing the minimum spanning tree of a connected, undirected graph `g` with optional
 distance matrix `distmx` using [Kruskal's algorithm](https://en.wikipedia.org/wiki/Kruskal%27s_algorithm).
 """
+
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: no space between docstring and function declaration.

File: src/spanningtrees/kruskal.jl [hunk 4]
```diff
@@ -39,22 +52,26 @@ function kruskal_mst end
...
             push!(mst, heap_entry.edge)
+            if length(mst) >= nv(g) - 1
+                break
+            end
         end
     end
 
     return mst
-end
+end
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: needs a blank line at the end.

File: src/spanningtrees/kruskal.jl (hunk 1 above)
Comment (SohamTamba) [reply src/spanningtrees/kruskal.jl]: Each set (connected component) is represented by a tree in a forest. The root of each tree represents the set. The while loop is necessary to is to traverse up the tree of the selected node q
and find the root representing the set q belongs to.

Additional [reference](https://ocw.mit.edu/courses/electrical-engineering-and-computer-science/6-046j-design-and-analysis-of-algorithms-spring-2015/recitation-notes/MIT6_046JS15_Recitation3.pdf) for the algorithm used

**Comment by SohamTamba** [comment]
I've made the suggested changes and squashed to one commit.

File: src/spanningtrees/kruskal.jl (hunk 1 above)
Comment (sbromberger) [reply src/spanningtrees/kruskal.jl]: I’m confused. The while loop checks for a condition, and then causes the condition to break the loop after the first iteration. root becomes set_id[root] in the first loop, so you will break the while immediately after. This is equivalent to an if.

Comment (sbromberger) [reply src/spanningtrees/kruskal.jl]: Ah, got it. Thanks.

**Comment by sbromberger** [comment]
Thank you for this, @SohamTamba !

**Pull request merged**